### PR TITLE
VLM: fixes after refactor

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

After merging https://github.com/huggingface/transformers/pull/30962 I noticed that some models stopped working in multi-image/image-video setting. Also `cache_position` wasn't being updated accordingly after merging image and text inputs in legacy way. This PR addresses the issue and adds slow tests for that

Also, I think we should use `num_image_tokens` instead of trying to infer it from image size and patch size. The reason is that some image backbones add CLS (CLIP) and other don't (SigLIP). If we try to catch every edge case like this, we'll end up bloating the code, so it's better to let users set the number of tokens needed for one image after encoding with the vision tower. That way we can support different vision towers out-of-the-box

Ran slow tests to test if expanding matches legacy, the new tests and a few older tests. Also changes video llava tests as those weren't formatting the prompt correctly

### CODE REVIEW

File: src/transformers/models/llava/processing_llava.py [hunk 1]
```diff
@@ -52,21 +52,21 @@ class LlavaProcessor(ProcessorMixin):
...
-    valid_kwargs = ["chat_template", "patch_size", "vision_feature_select_strategy", "image_token"]
+    valid_kwargs = ["chat_template", "num_image_tokens", "vision_feature_select_strategy", "image_token"]
     image_processor_class = "AutoImageProcessor"
     tokenizer_class = "AutoTokenizer"
 
     def __init__(
         self,
         image_processor=None,
         tokenizer=None,
-        patch_size=None,
```
Comment (amyeroberts) [inline src/transformers/models/llava/processing_llava.py:63]: Won't this thrown an error if this argument is removed, was previously part of a saved processors config and now isn't a valid kwarg?

File: src/transformers/models/llava/processing_llava.py [hunk 2]
```diff
@@ -52,21 +52,21 @@ class LlavaProcessor(ProcessorMixin):
...
+        num_image_tokens=None,
         vision_feature_select_strategy=None,
         chat_template=None,
         image_token="<image>",  # set the default and let users change if they have peculiar special tokens in rare cases
         **kwargs,
     ):
-        self.patch_size = patch_size
```
Comment (amyeroberts) [inline src/transformers/models/llava/processing_llava.py:69]: For BC we need to deprecate this attribute -- we can do this through creating a property we remove in a few cycles

File: src/transformers/models/llava_next/processing_llava_next.py [hunk 3]
```diff
@@ -183,13 +183,13 @@ def _get_number_of_features(self, orig_height: int, orig_width: int, height: int
         )
         scale_height, scale_width = height_best_resolution // height, width_best_resolution // width
 
-        patches_height = height // self.patch_size
-        patches_width = width // self.patch_size
+        patches_height = int(math.sqrt(self.num_image_tokens))
+        patches_width = int(math.sqrt(self.num_image_tokens))
```
Comment (amyeroberts) [inline src/transformers/models/llava_next/processing_llava_next.py:187]: This now assumes the image is square

File: tests/models/video_llava/test_modeling_video_llava.py
Comment (amyeroberts) [inline tests/models/video_llava/test_modeling_video_llava.py:1]: Why have all the prompts been changes to include `\n` in them?

**Review by amyeroberts — COMMENTED** [review]
This is a very big PR that touches a lot of modeling code in different ways. I just started reviewing but I think we should split up the changes here: switching to `num_image_tokens` is independent of fixing tests. This will make it easier to make sure the changes here are sensible. 

A few comments: 
* Why weren't the breaks caught previously? Was it lack of tests or the test fetcher? If lack of tests, tests should be added which will catch this
* I'm not sure about this switch to `num_image_tokens` -- can we guarantee this is always fixed? We need to properly handle this change in a non-breaking way
* Slow tests should be run for all these models

File: src/transformers/models/llava/processing_llava.py (hunk 1 above)
Comment (zucchini-nlp) [reply src/transformers/models/llava/processing_llava.py:63]: no, the prev checkpoints were not touched yet and afaik the new logic isn't used anywhere

File: tests/models/video_llava/test_modeling_video_llava.py
Comment (zucchini-nlp) [reply tests/models/video_llava/test_modeling_video_llava.py:1]: oh, that was a bad test from the beginning. Video-LLaVA expects the newline char in its prompt, it wasn't formatted correctly for the tests

**Comment by zucchini-nlp** [comment]
Oke, I will separate these into two PRs and then I have another one, which adds Generation tests in VLMs. That will be a third PR

- Why weren't the breaks caught previously? Was it lack of tests or the test fetcher? If lack of tests, tests should be added which will catch this
Yes, no tests for these cases were added, I added some more. I'll add one more for llava-interleave checkpoint as it uses SigLIP as backbone

- I'm not sure about this switch to num_image_tokens -- can we guarantee this is always fixed? We need to properly handle this change in a non-breaking way
If we didn't change any checkpoint and the latest major release still doesn't have the new processing logic, this shouldn't cause BC errors. WDYT? 

 - Slow tests should be run for all these models
 Yes, will enable in each new PR

**Comment by zucchini-nlp** [comment]
To be reviewed and merged after https://github.com/huggingface/transformers/pull/33168

**Comment by amyeroberts** [comment]
@zucchini-nlp OK - just ping when you want another review!

File: src/transformers/models/llava_next/processing_llava_next.py [hunk 4]
```diff
@@ -140,30 +140,29 @@ def __call__(
...
-                sample = sample.replace(self.image_token, self.image_token * num_image_tokens)
-                prompt_strings.append(sample)
+        prompt_strings = text
+        if image_inputs:
+            if self.patch_size is None or self.vision_feature_select_strategy is None:
+                logger.warning_once(
+                    "Expanding inputs for image tokens in LLaVa-NeXT should be done in processing. "
+                    "Please add `patch_size` and `vision_feature_select_strategy` to the model's processing config or set directly "
+                    "with `processor.patch_size = {{patch_size}}` and processor.vision_feature_select_strategy = {{vision_feature_select_strategy}}`. "
+                    "Using processors without these attributes in the config is deprecated and will throw an error in v4.47."
```
Comment (zucchini-nlp) [inline src/transformers/models/llava_next/processing_llava_next.py:150]: This was needed for cases with mutli-image inputs where we cannot be sure that number of image sizes is same as text. For ex, one text and two images

File: src/transformers/models/llava_next/processing_llava_next.py [hunk 5]
```diff
@@ -199,8 +198,8 @@ def _get_unpadded_features(self, height, width, patches_height, patches_width, s
         because it divided each image into patches depending on its resolution. Therefore we need to calculate how many
         patches an image is divided into and get the number of features from that.
         """
-        current_width = patches_height * scale_height
-        current_height = patches_width * scale_width
+        current_height = patches_height * scale_height
+        current_width = patches_width * scale_width
```
Comment (zucchini-nlp) [inline src/transformers/models/llava_next/processing_llava_next.py:202]: The order was corrected in other PR for modeling code but I forgot to add same correction to the processing logic

File: src/transformers/models/llava_next_video/diff_llava_next_video.py [hunk 6]
```diff
@@ -389,13 +388,17 @@ def forward(
...
-            inputs_expanded = (
-                img_token_count < self.config.image_seq_length and video_token_count < self.config.video_seq_length
+            img_token_not_enough = (input_ids == self.config.image_token_index).sum(
+                1
+            ).max() < self.config.image_seq_length
+            video_token_not_enough = (input_ids == self.config.video_token_index).sum(
+                1
+            ).max() < self.config.video_seq_length
+            inputs_not_expanded = (img_token_not_enough and pixel_values is not None) or (
+                video_token_not_enough and pixel_values_videos is not None
```
Comment (zucchini-nlp) [inline src/transformers/models/llava_next_video/diff_llava_next_video.py:398]: better check for multiple visual per prompt

File: src/transformers/models/llava_next_video/processing_llava_next_video.py [hunk 7]
```diff
@@ -197,23 +192,65 @@ def __call__(
...
             truncation=truncation,
             max_length=max_length,
         )
-        print(text_inputs.keys())
-
         return BatchFeature(data={**text_inputs, **image_inputs, **videos_inputs})
 
+    # Copied from transformers.models.llava_next.processing_llava_next.LlavaNextProcessor._get_number_of_features
+    def _get_number_of_features(self, orig_height: int, orig_width: int, height: int, width: int) -> int:
+        image_grid_pinpoints = self.image_processor.image_grid_pinpoints
```
Comment (zucchini-nlp) [inline src/transformers/models/llava_next_video/processing_llava_next_video.py:212]: we forgot about image inputs for this model, added a test

**Comment by zucchini-nlp** [comment]
@amyeroberts this is ready for review now, the blocking PR is merged. For the tests, the failing ones are SDPA-related which is fixed in https://github.com/huggingface/transformers/pull/32238 (needs a separate review, but not a blocker for current PR). 

The tests are fixed so that we match the runner's output. I verified that the discrepancy appears after updating torch to the latest version and the test was passing in `2.3.0`

File: src/transformers/models/llava_next/processing_llava_next.py [hunk 8]
```diff
@@ -140,30 +140,29 @@ def __call__(
...
+                    "with `processor.patch_size = {{patch_size}}` and processor.vision_feature_select_strategy = {{vision_feature_select_strategy}}`. "
+                    "Using processors without these attributes in the config is deprecated and will throw an error in v4.47."
+                )
+            else:
+                image_sizes = iter(image_inputs["image_sizes"])
+                height, width = get_image_size(to_numpy_array(image_inputs["pixel_values"][0][0]))
+                prompt_strings = []
+                for sample in text:
+                    while self.image_token in sample:
+                        image_size = next(image_sizes)
```
Comment (amyeroberts) [inline src/transformers/models/llava_next/processing_llava_next.py:158]: Is this right? The previous logic implies `len(image_sizes) == len(text)` however this is implying we have the same number of image_sizes as the number of image tokens per sample * number of samples.

**Review by amyeroberts — COMMENTED** [review]
Thanks for working on this fix! 

There's just one outstanding question I have about the iterator logic for image_sizes

File: tests/models/vipllava/test_modeling_vipllava.py [hunk 9]
```diff
@@ -271,26 +271,23 @@ def test_small_model_integration_test(self):
...
         pixel_values = torch.randn(
-            (2, 3, 336, 336),
+            (1, 3, 336, 336),
             dtype=torch.float,
             device=torch_device,
         )
         input_ids = torch.tensor(
             [
                 [32001, 32001, 1, 15043, 7084, 32000, 29871, 13, 7900],
-                [1, 15043, 7084, 29901, 29871, 32000, 29871, 13, 7900],
```
[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply tests/models/vipllava/test_modeling_vipllava.py:287]: yep, to save up memory as this is a test with grad and a backward call, so it needs more memory than generation

File: src/transformers/models/llava_next/processing_llava_next.py (hunk 8 above)
Comment (zucchini-nlp) [reply src/transformers/models/llava_next/processing_llava_next.py:158]: Yes, I verified twice to be sure. The number of images == number of image sizes. The new added test `multiimage_expansion` fails with the previous logic

Comment (zucchini-nlp) [reply src/transformers/models/llava_next/processing_llava_next.py:158]: @amyeroberts do you have any other concerns regarding this PR?

**Review by amyeroberts — APPROVED** [review]
Thanks for double checking the logic!

**Pull request merged**